### PR TITLE
Halloween Lint using Putout (part 5)

### PR DESCRIPTION
**Pull request by coderaiser**
After successfully merged linting sessions [part 1](https://github.com/xtermjs/xterm.js/pull/2953), [part 2](https://github.com/xtermjs/xterm.js/pull/3133) and [part 3](https://github.com/xtermjs/xterm.js/pull/3269), [part 4](https://github.com/xtermjs/xterm.js/pull/3341) the time is come for part 5 :).

As usual, any rule can be disabled.

Command used:

```sh
putout . --fix
```

Applied rules:
- [regexp/optimize](https://github.com/coderaiser/putout/tree/v21.1.0/packages/plugin-regexp);
- [apply-optional-chaining](https://github.com/coderaiser/putout/tree/v21.1.0/packages/plugin-apply-optional-chaining);
- [convert-index-of-to-includes](https://github.com/coderaiser/putout/tree/v21.1.0/packages/plugin-convert-index-of-to-includes);
- [convert-array-copy-to-slice](https://github.com/coderaiser/putout/tree/v21.1.0/packages/plugin-convert-array-copy-to-slice);
- [remove-useless-escape](https://github.com/coderaiser/putout/tree/v21.1.0/packages/plugin-remove-useless-escape);
- [remove-unused-expressions](https://github.com/coderaiser/putout/tree/v21.1.0/packages/remove-unused-expressions);
- [remove-useless-operand](https://github.com/coderaiser/putout/tree/v21.1.0/packages/plugin-remove-useless-operand);
- [apply-is-array](https://github.com/coderaiser/putout/tree/v21.1.0/packages/plugin-apply-is-array);

Current config for [putout v21](https://github.com/coderaiser/putout):

```json
{
    "rules": {
        "apply-is-array": ["on", {
            "inline": true
        },
        "remove-useless-new": "off",
        "remove-useless-return": "off",
        "remove-useless-types-from-constants": "off",
        "remove-useless-type-convertion/with-double-negations": "off",
        "convert-typeof-to-is-type": "off",
        "apply-shorthand-properties": "off",
        "apply-destructuring": "off",
        "apply-numeric-separators": "off",
        "convert-assignment-to-comparison": "off",
        "convert-apply-to-spread": "off",
        "convert-math-pow": "off",
        "convert-for-to-for-of": "off",
        "convert-template-to-string": "off",
        "strict-mode": "off",
        "remove-useless-spread/object": "off",
        "remove-useless-array-constructor": "off",
        "remove-boolean-from-assertions": "off",
        "remove-iife": "off",
        "remove-console": "off",
        "remove-unused-variables": "off",
        "remove-useless-variables": "off",
        "merge-if-statements": "off",
        "promises/add-missing-await": "off",
        "promises/remove-useless-async": "off",
        "simplify-ternary": "off"
    },
    "match": {
        "*.benchmark.ts": {
            "remove-unused-expressions": "off"
        }
    },
    "plugins": [
        "apply-shorthand-properties"
    ],
    "ignore": [
        "*.md",
        "*.json",
        "*ignore",
        "*.yml",
        ".npmrc",
        "*.css",
        "out",
        "out-test",
        "demo"
    ]
}
```

Happy Halloween 🎃!

### CODE REVIEW

**Review by Tyriar — CHANGES_REQUESTED** [review]
Looks like this caused tsc strict errors:

```
src/common/InputHandler.ts(64,18): error TS7006: Parameter 'a' implicitly has an 'any' type.
src/common/InputHandler.ts(561,45): error TS2339: Property 'substring' does not exist on type 'string | Uint8Array'.
  Property 'substring' does not exist on type 'Uint8Array'.
src/common/InputHandler.ts(562,43): error TS2339: Property 'subarray' does not exist on type 'string | Uint8Array'.
  Property 'subarray' does not exist on type 'string'.
src/common/input/WriteBuffer.test.ts(12,18): error TS7006: Parameter 'a' implicitly has an 'any' type.
src/common/input/WriteBuffer.test.ts(49,75): error TS2345: Argument of type 'string | Uint8Array' is not assignable to parameter of type 'Uint8Array'.
  Type 'string' is not assignable to type 'Uint8Array'.
src/common/parser/EscapeSequenceParser.test.ts(17,18): error TS7006: Parameter 'a' implicitly has an 'any' type.
src/common/parser/EscapeSequenceParser.test.ts(1737,7): error TS2322: Type 'unknown' is not assignable to type 'string | undefined'.
  Type 'unknown' is not assignable to type 'string'.
addons/xterm-addon-attach/src/AttachAddon.ts(10,18): error TS7006: Parameter 'a' implicitly has an 'any' type.
addons/xterm-addon-attach/src/AttachAddon.ts(32,24): error TS2345: Argument of type 'string | ArrayBuffer' is not assignable to parameter of type 'string | Uint8Array'.
  Type 'ArrayBuffer' is not assignable to type 'string | Uint8Array'.
    Type 'ArrayBuffer' is missing the following properties from type 'Uint8Array': BYTES_PER_ELEMENT, buffer, byteOffset, copyWithin, and 22 more.
addons/xterm-addon-attach/src/AttachAddon.ts(32,63): error TS2769: No overload matches this call.
  The last overload gave the following error.
    Argument of type 'string | ArrayBuffer' is not assignable to parameter of type 'ArrayBuffer'.
      Type 'string' is not assignable to type 'ArrayBuffer'.
error TS2688: Cannot find type definition file for '../../../out-test/api/TestUtils'.
  The file is in the program because:
    Entry point of type library '../../../out-test/api/TestUtils' specified in compilerOptions
error TS2688: Cannot find type definition file for '../../../out-test/api/TestUtils'.
  The file is in the program because:
    Entry point of type library '../../../out-test/api/TestUtils' specified in compilerOptions
error TS2688: Cannot find type definition file for '../../../out-test/api/TestUtils'.
  The file is in the program because:
    Entry point of type library '../../../out-test/api/TestUtils' specified in compilerOptions
error TS2688: Cannot find type definition file for '../../../out-test/api/TestUtils'.
  The file is in the program because:
    Entry point of type library '../../../out-test/api/TestUtils' specified in compilerOptions
```

**Comment by coderaiser** [comment]
Disabled:
-  [`remove-useless-types-from-constants`](https://github.com/coderaiser/putout/tree/master/packages/plugin-remove-useless-types-from-constants);
- [`convert-typeof-to-is-type`](https://github.com/coderaiser/putout/tree/master/packages/plugin-convert-typeof-to-is-type);

**Comment by jerch** [comment]
@coderaiser

On a sidenote: most promises removed from tests, are actually needed to force the return of a promise testing promise handling in the parser. So they should not be removed there.

**Comment by coderaiser** [comment]
@jerch just got back promises

**Comment by coderaiser** [comment]
Disabled:
- [remove-useless-type-conversion/with-double-negations](https://github.com/coderaiser/putout/tree/v21.1.0/packages/plugin-remove-useless-type-conversion);

File: addons/xterm-addon-serialize/benchmark/SerializeAddon.benchmark.ts [hunk 1]
```diff
@@ -58,8 +58,8 @@ perfContext('Terminal: sh -c "dd if=/dev/urandom count=40 bs=1k | hexdump | lolc
       serializeAddon.activate(terminal);
       terminal.writeSync(content);
     });
-    new ThroughputRuntimeCase('', () => {
-      return { payloadSize: serializeAddon.serialize().length };
-    }, { fork: false }).showAverageThroughput();
+    new ThroughputRuntimeCase('', () => (({
+      payloadSize: serializeAddon.serialize().length
+    })), { fork: false }).showAverageThroughput();
```
Comment (Tyriar) [inline addons/xterm-addon-serialize/benchmark/SerializeAddon.benchmark.ts:63]: This was more readable before imo, also it doubled up the params for some reasons.

File: addons/xterm-addon-ligatures/src/font.ts [hunk 2]
```diff
@@ -41,7 +41,7 @@ export default async function load(fontFamily: string, cacheSize: number): Promi
           name: 'local-fonts'
         });
         if (status && status.state !== 'granted') {
-          throw new Error('Permission to access local fonts not granted.');
+          throw Error('Permission to access local fonts not granted.');
```
Comment (Tyriar) [inline addons/xterm-addon-ligatures/src/font.ts:44]: I kind of prefer `new Error` and `new Array`, let's remove that rule as well so this PR becomes a lot smaller. This is something a compiler should remove if they are equivalent imo.

File: src/common/CircularList.ts [hunk 3]
```diff
@@ -52,7 +52,7 @@ export class CircularList<T> implements ICircularList<T> {
 
     // Reconstruct array, starting at index 0. Only transfer values from the
     // indexes 0 to length.
-    const newArray = new Array<T | undefined>(newMaxLength);
+    const newArray = Array(newMaxLength);
```
Comment (Tyriar) [inline src/common/CircularList.ts:55]: Not sure this was a safe thing to do removing the type here?

File: src/common/parser/Params.ts [hunk 4]
```diff
@@ -9,6 +9,10 @@ const MAX_VALUE = 0x7FFFFFFF;
 // max allowed subparams for a single sequence (hardcoded limitation)
 const MAX_SUBPARAMS = 256;
 
+const {
+  isArray
+} = Array;
```
Comment (Tyriar) [inline src/common/parser/Params.ts:14]: Can we inline `Array.isArray` instead?

**Review by Tyriar — CHANGES_REQUESTED** [review]
Lots of good changes here, I especially like the optional chaining.

**Comment by coderaiser** [comment]
Disabled:
- [`remove-useless-new`](https://github.com/coderaiser/putout/tree/v21.1.0/packages/plugin-remove-useless-new);
- [`remove-useless-return`](https://github.com/coderaiser/putout/tree/v21.1.0/packages/plugin-remove-useless-return);

**Comment by coderaiser** [comment]
Disabled:
- [`remove-useless-new`](https://github.com/coderaiser/putout/tree/v21.1.0/packages/plugin-remove-useless-new);
- [`remove-useless-return`](https://github.com/coderaiser/putout/tree/v21.1.0/packages/plugin-remove-useless-return);

**Comment by coderaiser** [comment]
Disabled:
- [`remove-useless-new`](https://github.com/coderaiser/putout/tree/v21.1.0/packages/plugin-remove-useless-new);
- [`remove-useless-return`](https://github.com/coderaiser/putout/tree/v21.1.0/packages/plugin-remove-useless-return);

File: src/common/parser/Params.ts [hunk 5]
```diff
@@ -9,6 +9,7 @@ const MAX_VALUE = 0x7FFFFFFF;
 // max allowed subparams for a single sequence (hardcoded limitation)
 const MAX_SUBPARAMS = 256;
 
+
```
Comment (coderaiser) [inline src/common/parser/Params.ts:12]:
```suggestion
```

File: src/common/parser/Params.ts (hunk 4 above)
Comment (coderaiser) [reply src/common/parser/Params.ts]: Landed `inline` option to [`@putout/plugin-apply-is-array`](https://github.com/coderaiser/putout/tree/master/packages/plugin-apply-is-array#-correct-code-example) `v2.1.0`

**Review by Tyriar — APPROVED** [review]
Thanks, always appreciated 🙂

**Pull request merged**